subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java

```diff
@@ -66,24 +68,32 @@ public DefaultCppBinary(Names names, ObjectFactory objects, Provider<String> bas
 
         // TODO - reduce duplication with Swift binary
-
         includePathConfiguration = configurations.create(names.withPrefix("cppCompile"));
         includePathConfiguration.setCanBeConsumed(false);
-        includePathConfiguration.getAttributes().attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.C_PLUS_PLUS_API));
-        includePathConfiguration.getAttributes().attribute(DEBUGGABLE_ATTRIBUTE, identity.isDebuggable());
-        includePathConfiguration.getAttributes().attribute(OPTIMIZED_ATTRIBUTE, identity.isOptimized());
-        includePathConfiguration.getAttributes().attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, identity.getTargetMachine().getOperatingSystemFamily());
-        includePathConfiguration.getAttributes().attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, identity.getTargetMachine().getArchitecture());
-        includePathConfiguration.extendsFrom(getImplementationDependencies());
-
-        Configuration nativeLink = configurations.create(names.withPrefix("nativeLink"));
-        nativeLink.setCanBeConsumed(false);
-        nativeLink.getAttributes().attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.NATIVE_LINK));
-        nativeLink.getAttributes().attribute(DEBUGGABLE_ATTRIBUTE, identity.isDebuggable());
-        nativeLink.getAttributes().attribute(OPTIMIZED_ATTRIBUTE, identity.isOptimized());
-        nativeLink.getAttributes().attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, identity.getTargetMachine().getOperatingSystemFamily());
-        nativeLink.getAttributes().attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, identity.getTargetMachine().getArchitecture());
-        nativeLink.extendsFrom(getImplementationDependencies());
+        includePathConfiguration.setCanBeResolved(false);
+
+        includePathPathConfiguration = configurations.create(names.withPrefix("cppCompilePath"));
+        includePathPathConfiguration.setCanBeConsumed(false);
+        includePathPathConfiguration.getAttributes().attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.C_PLUS_PLUS_API));
+        includePathPathConfiguration.getAttributes().attribute(DEBUGGABLE_ATTRIBUTE, identity.isDebuggable());
+        includePathPathConfiguration.getAttributes().attribute(OPTIMIZED_ATTRIBUTE, identity.isOptimized());
+        includePathPathConfiguration.getAttributes().attribute(OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE, identity.getTargetMachine().getOperatingSystemFamily());
+        includePathPathConfiguration.getAttributes().attribute(MachineArchitecture.ARCHITECTURE_ATTRIBUTE, identity.getTargetMachine().getArchitecture());
+        includePathPathConfiguration.extendsFrom(includePathConfiguration, getImplementationDependencies());
+
+        linkLibraries = configurations.create(names.withPrefix("nativeLink"));
+        linkLibraries.setVisible(false);
+        linkLibraries.setCanBeConsumed(false);
+        linkLibraries.setCanBeResolved(false);
```

Comment:
❌  Why are any of these changes required? I thought we were already using the "correct" way of dealing with variant-aware dependencies. It seems we are using a lot more layers here. The issue seems that we are using the `nativeLink` and `cppCompile` for consumption and bucket of dependencies (case 5).

I don't think the new bucket name (with `Path` suffix) make much sense for native. It should be aligned with the JVM domain. The name would make more sense as `compileOnly`, `linkOnly` and `runtimeOnly`. However, we don't have tests for those buckets and is not even a thing in native yet. It was just coincidental that it worked. I would lean toward leaving it as it is, showing a message deprecating message and working toward a better solution for pre-built binaries.

My understanding is if we leave it as is, there will be deprecation warning issued. I think it's fine for now, we can revisit those as a separate issue.

Reply:
Thanks @lacasseio for taking the time. Your understanding is correct and what you suggest makes sense to me. I'll remove the change from this PR and open a separate issue.

Reply:
FYI. These are the tests that trigger a deprecation warning without the change: https://builds.gradle.org/viewLog.html?buildId=23768726